File: answers/leetcode/3Sum Closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        nums.sort()
        min_diff = sys.maxint
        three_sum = 0
        for i in range(len(nums)):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                diff = abs(sum - target)
                if min_diff > diff:
                    min_diff = diff
                    three_sum = sum
                if sum > target:
                    right -= 1
                elif sum < target:
                    left += 1
                else:
                    return three_sum
        return three_sum

File: answers/leetcode/3Sum Closest/3Sum Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int result = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < nums.length; i++) {
            int left = i + 1;
            int right = nums.length - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (Math.abs(sum - target) < Math.abs(result - target)) {
                    result = sum;
                }
                if (sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                } else {
                    break;
                }
            }
        }

        return result;
    }
}
//Runtime: 14 ms, faster than 17.20% of Java online submissions for 3Sum Closest.
//Memory Usage: 39.8 MB, less than 21.23% of Java online submissions for 3Sum Closest.
